core/commands/ls.go

```diff
@@ -173,45 +155,37 @@ The JSON output contains type information.
 			}
 		}
 
-		res.SetOutput(&LsOutput{output})
+		return res.Emit(&LsOutput{output})
 	},
-	Marshalers: cmds.MarshalerMap{
-		cmds.Text: func(res cmds.Response) (io.Reader, error) {
-
-			v, err := unwrapOutput(res.Output())
-			if err != nil {
-				return nil, err
-			}
-
-			headers, _, _ := res.Request().Option(lsHeadersOptionNameTime).Bool()
-			output, ok := v.(*LsOutput)
-			if !ok {
-				return nil, e.TypeErr(output, v)
-			}
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out *LsOutput) error {
+			headers, _ := req.Options[lsHeadersOptionNameTime].(bool)
 
 			buf := new(bytes.Buffer)
-			w := tabwriter.NewWriter(buf, 1, 2, 1, ' ', 0)
-			for _, object := range output.Objects {
-				if len(output.Objects) > 1 {
-					fmt.Fprintf(w, "%s:\n", object.Hash)
+			tw := tabwriter.NewWriter(buf, 1, 2, 1, ' ', 0)
```

Comment:
Let's pass `w` into `tabwriter.NewWriter` directly (and get rid of `buf` entirely). We used to put this all in a buffer because we had to return it from the command but that's no longer necessary.